### ISSUE TITLE
logic bug in dynamic template substitution

### ISSUE THREAD

**Issue by codmajik** [sendgrid/sendgrid-nodejs#747]
#### Issue Summary
Passing dynamic or substitution wont work for dynamic templates because of the logic bug found here

https://github.com/sendgrid/sendgrid-nodejs/blob/0c83ae9a7151f466383df35a897a3426c5cdb06b/packages/helpers/classes/mail.js#L243-L247

after 2 days trying to find why substitution is failing I decided to look in the code.

The current work-around is to supply both `substitutions` and `dynamicTemplateData` in which case the library will delete `substitutions`, leaving `dynamicTemplateData`

**Comment by thinkingserious** [comment]
Hello @codmajik,

Thanks for reporting this! I've added this to our backlog for further investigation.

With Best Regards,

Elmer

**Comment by mateoholman** [comment]
Hello, I'd love to take on this one if no one is currently working on it?

**Comment by thinkingserious** [comment]
It's all yours @mateoholman, thanks!

**Comment by ajsharp** [comment]
@thinkingserious I'd love to get this fixed, but I need a bit of guidance on how the library _should_ behave.

Can you summarize what the behavior here should be? It seems like the logic in question is trying to prevent the a user of this library from using the wrong property in the wrong context (`substitutions` vs `dynamicTemplateData`), as opposed to just sending the data along that the user has passed in.

It seems like the intended logic is, if the user is using a dynamic template (a template id that starts with a 'd'), then they should be using the `dynamicTemplateData` property, otherwise, for legacy templates, they should be using `substitutions`. It's not clear to me in which contexts the dynamicTemplateData property is supposed to be nested inside `personalizations`, but the library appears to expect it in both forms.

I'm not sure this library needs this logic, which seems intended to prevent the user from making a logical mistake, which presumably the api will respond with an error anyway. As a general principle, I don't think this library should be deleting or materially changing the structure of data the user has told the library to pass along to the api. That's unexpected and leads to very difficult to track down bugs.

Also, it seems that there is a related bug here in the typescript definition file that only accepts the `dynamicTemplateData` property inside the `personalizations`, but not at the top level, which seems to be what the javascript implementation expects: https://github.com/sendgrid/sendgrid-nodejs/blob/95dfe0652abb29eeae41b4530f12d5d7be95d3ed/packages/helpers/classes/mail.js#L90-L92.

I'm happy to fix this in a PR, I'm just not sure what the behavior here should be.

**Comment by ajsharp** [comment]
@thinkingserious it looks like the typescript issue has been resolved here: https://github.com/sendgrid/sendgrid-nodejs/commit/2d55dc44b6028a6bcc2e26400ce187a905a8c754. Any plans to update the package with this change?

**Comment by ajsharp** [comment]
@thinkingserious ping

**Comment by thinkingserious** [comment]
@mateoholman, @ajsharp,

Are either of you working on this one?

@ajsharp,

I'm hoping to get a release out soon (Node.js is next in our queue), but I don't have a specific date.

**Comment by childish-sambino** [comment]
Fixed by #939
